lib/kube/proxy/transport.go

```diff
@@ -0,0 +1,489 @@
+		// handshakes performed.
+		MaxIdleConns:        defaults.HTTPMaxIdleConns,
+		MaxIdleConnsPerHost: defaults.HTTPMaxIdleConnsPerHost,
+		// IdleConnTimeout defines the maximum amount of time before idle connections
+		// are closed. Leaving this unset will lead to connections open forever and
+		// will cause memory leaks in a long running process.
+		IdleConnTimeout: defaults.HTTPIdleTimeout,
+	}
+}
+
+// versionWithoutImpersonation is the version of Teleport that starts supporting
+// impersonation. Before this version, the client will not use impersonation.
+var versionWithoutImpersonation = semver.New(utils.VersionBeforeAlpha("13.0.0"))
+
+// teleportVersionInterface is an interface that allows to get the Teleport version of
+// a server.
+// DELETE IN 14.0.0
+type teleportVersionInterface interface {
+	GetTeleportVersion() string
+}
+
+// allServersSupportImpersonation returns true if all servers in the list
+// support impersonation. This is used to determine if the client should
+// create a new client certificate and use a different [http.Transport]
+// (https://golang.org/pkg/net/http/#Transport) for each request.
+// Only returns true if all servers in the list support impersonation.
+// DELETE IN 14.0.0
+func allServersSupportImpersonation[T teleportVersionInterface](servers []T) bool {
+	if len(servers) == 0 {
+		return false
+	}
+	for _, server := range servers {
+		serverVersion := server.GetTeleportVersion()
+		semVer, err := semver.NewVersion(serverVersion)
+		if err != nil || semVer.LessThan(*versionWithoutImpersonation) {
+			return false
+		}
+	}
+	return true
+}
+
+// getOrRequestClientCreds returns the client credentials for the provided auth context.
+// If the credentials are not cached, they will be requested from the auth server.
+// DELETE IN 14.0.0
+func (f *Forwarder) getOrRequestClientCreds(tracingCtx context.Context, authCtx authContext) (*tls.Config, error) {
+	c := f.getClientCreds(authCtx)
+	if c == nil {
+		return f.serializedRequestClientCreds(tracingCtx, authCtx)
+	}
+	return c, nil
+}
+
+// getClientCreds returns the client credentials for the provided auth context.
+// DELETE IN 14.0.0
+func (f *Forwarder) getClientCreds(ctx authContext) *tls.Config {
+	f.mu.Lock()
+	defer f.mu.Unlock()
+	creds, ok := f.clientCredentials.Get(ctx.key())
+	if !ok {
+		return nil
+	}
+	c := creds.(*tls.Config)
+	if !validClientCreds(f.cfg.Clock, c) {
+		return nil
+	}
+	return c
+}
+
+// saveClientCreds saves the client credentials for the provided auth context.
+// DELETE IN 14.0.0
+func (f *Forwarder) saveClientCreds(ctx authContext, c *tls.Config) error {
+	f.mu.Lock()
+	defer f.mu.Unlock()
+	return f.clientCredentials.Set(ctx.key(), c, ctx.sessionTTL)
+}
+
+// validClientCreds returns true if the provided client credentials are valid.
+// DELETE IN 14.0.0
+func validClientCreds(clock clockwork.Clock, c *tls.Config) bool {
+	if len(c.Certificates) == 0 || len(c.Certificates[0].Certificate) == 0 {
```

Comment:
Don't we use the pattern of replacing a singleton cert in `c.Certificates` with a `GetClientCertificate` closure a lot?

Reply:
I was unware of it. I just moved that code and I really expect to remove it for the next teleport release.
Let me know if you want me to change it

Reply:
Didn't realize that it was just moved code - if it worked fine so far it's probably ok. 🤔